distributed/diagnostics/plugin.py

```diff
@@ -282,34 +278,74 @@ def __init__(self, packages, pip_options=None, restart=False):
         self.packages = packages
         self.restart = restart
         self.pip_options = pip_options or []
+        self.id = f"pip-install-{uuid.uuid4()}"
 
     async def setup(self, worker):
-        from distributed.lock import Lock
-
-        async with Lock(socket.gethostname()):  # don't clobber one installation
-            logger.info("Pip installing the following packages: %s", self.packages)
-            proc = subprocess.Popen(
-                [sys.executable, "-m", "pip", "install"]
-                + self.pip_options
-                + self.packages,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
-            )
-            stdout, stderr = proc.communicate()
-            returncode = proc.wait()
-
-            if returncode:
-                logger.error("Pip install failed with '%s'", stderr.decode().strip())
-                return
-
-            if self.restart and worker.nanny:
-                lines = stdout.strip().split(b"\n")
-                if not all(
-                    line.startswith(b"Requirement already satisfied") for line in lines
-                ):
-                    worker.loop.add_callback(
-                        worker.close_gracefully, restart=True
-                    )  # restart
+        from distributed.semaphore import Semaphore
+
+        async with (
+            await Semaphore(max_leases=1, name=socket.gethostname(), register=True)
+        ):
+            if not await self._is_installed(worker):
+                logger.info("Pip installing the following packages: %s", self.packages)
+                await self._set_installed(worker)
+                proc = subprocess.Popen(
+                    [sys.executable, "-m", "pip", "install"]
+                    + self.pip_options
+                    + self.packages,
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.PIPE,
+                )
+                stdout, stderr = proc.communicate()
+                returncode = proc.wait()
+                if returncode != 0:
+                    logger.error(
+                        "Pip install failed with '%s'", stderr.decode().strip()
+                    )
+                    return
```

Comment:
I think that in addition to logging the error, we may want to raise a `RuntimeError` here instead of returning. This way, we can propagate the error to the user.